tiledb/api/src/datatype/logical.rs

```diff
@@ -0,0 +1,143 @@
+use crate::datatype::physical::PhysicalType;
+use crate::datatype::Datatype;
+use crate::private::sealed;
+
+/// Trait which provides statically-typed attributes for a TileDB `Datatype`
+/// for use with generics.
+pub trait LogicalType: crate::private::Sealed {
+    const DATA_TYPE: Datatype;
+
+    type PhysicalType: PhysicalType;
+}
+
+pub struct UInt8Type {}
+
+impl LogicalType for UInt8Type {
+    const DATA_TYPE: Datatype = <u8 as PhysicalType>::DATA_TYPE;
```

Comment:
Not necessarily asking for a change here, but I will say this pattern took me more than 30s to figure out. On the one hand, it seems like just writing out `DataType::UInt8` would be clearer, though on the other hand, there's a vague "But then it's possible to get things mismatched". Personally, I'd lean towards "this is write once code that'll ~never change, only possibly be replaced wholesale" so I'd go with the `Datatype::$variant` to make things clearer.

Granted, that's only a very loosely held slight preference so feel free to ignore this.

Reply:
I was kind of unsure if the `PhysicalType` trait should have a Datatype associated constant at all, so maybe I'll remove that and follow along with this.

Reply:
Fair point. I could maybe see there being a "default logical type", though given my assumption that most uses of our API bindings will be "programmatic" instead of user written so the ergonomics probably won't matter there.